In [12]:
import pandas as pd
import numpy as np
import re

In [241]:
df = pd.read_excel("C:/Users/NaEunSu/critical_shoot_project/02.team_project_data/서울시 공공자전거 이용현황/19.01~21.06 자전거 대여 현황.xlsx")
df

대여일자      대여건수  Unnamed: 2  Unnamed: 3  Unnamed: 4
0   2019-01-01    8601.0         NaN         NaN         NaN
1   2019-01-02   14360.0         NaN         NaN         NaN
2   2019-01-03   16344.0         NaN         NaN         NaN
3   2019-01-04   17308.0         NaN         NaN         NaN
4   2019-01-05   12380.0         NaN         NaN         NaN
..         ...       ...         ...         ...         ...
875 2021-06-26  105813.0         NaN         NaN         NaN
876 2021-06-27   92869.0         NaN         NaN         NaN
877 2021-06-28  106378.0         NaN         NaN         NaN
878 2021-06-29  111907.0         NaN         NaN         NaN
879 2021-06-30  122200.0         NaN         NaN         NaN

[880 rows x 5 columns]

In [242]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880 entries, 0 to 879
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   대여일자        880 non-null    datetime64[ns]
 1   대여건수        464 non-null    float64       
 2   Unnamed: 2  416 non-null    float64       
 3   Unnamed: 3  0 non-null      float64       
 4   Unnamed: 4  0 non-null      float64       
dtypes: datetime64[ns](1), float64(4)
memory usage: 34.5 KB


info 확인 결과 대여건수가 대여건수 컬럼에 464개, Unnamed: 2 컬럼에 416개로 나눠져 있었음

그래서 대여건수와 Unnamed: 2를 합쳐주는 과정을 진행하겠음

In [243]:
df["대여건수"] = df["대여건수"].fillna(0)

In [244]:
df["Unnamed: 2"] = df["Unnamed: 2"].fillna(0)

In [245]:
df["대여건수"] = df["대여건수"]+df["Unnamed: 2"]

In [246]:
df

대여일자      대여건수  Unnamed: 2  Unnamed: 3  Unnamed: 4
0   2019-01-01    8601.0         0.0         NaN         NaN
1   2019-01-02   14360.0         0.0         NaN         NaN
2   2019-01-03   16344.0         0.0         NaN         NaN
3   2019-01-04   17308.0         0.0         NaN         NaN
4   2019-01-05   12380.0         0.0         NaN         NaN
..         ...       ...         ...         ...         ...
875 2021-06-26  105813.0         0.0         NaN         NaN
876 2021-06-27   92869.0         0.0         NaN         NaN
877 2021-06-28  106378.0         0.0         NaN         NaN
878 2021-06-29  111907.0         0.0         NaN         NaN
879 2021-06-30  122200.0         0.0         NaN         NaN

[880 rows x 5 columns]

In [247]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880 entries, 0 to 879
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   대여일자        880 non-null    datetime64[ns]
 1   대여건수        880 non-null    float64       
 2   Unnamed: 2  880 non-null    float64       
 3   Unnamed: 3  0 non-null      float64       
 4   Unnamed: 4  0 non-null      float64       
dtypes: datetime64[ns](1), float64(4)
memory usage: 34.5 KB


대여건수에 880개로 합쳐져있는걸 확인

In [248]:
# 불피요한 컬럼 제거
df = df.iloc[ : , :2]
df

대여일자      대여건수
0   2019-01-01    8601.0
1   2019-01-02   14360.0
2   2019-01-03   16344.0
3   2019-01-04   17308.0
4   2019-01-05   12380.0
..         ...       ...
875 2021-06-26  105813.0
876 2021-06-27   92869.0
877 2021-06-28  106378.0
878 2021-06-29  111907.0
879 2021-06-30  122200.0

[880 rows x 2 columns]

In [251]:
# 대여일자를 행 인덱스로 설정
df.set_index(df['대여일자'], inplace=True)

In [252]:
df

대여일자      대여건수
대여일자                           
2019-01-01 2019-01-01    8601.0
2019-01-02 2019-01-02   14360.0
2019-01-03 2019-01-03   16344.0
2019-01-04 2019-01-04   17308.0
2019-01-05 2019-01-05   12380.0
...               ...       ...
2021-06-26 2021-06-26  105813.0
2021-06-27 2021-06-27   92869.0
2021-06-28 2021-06-28  106378.0
2021-06-29 2021-06-29  111907.0
2021-06-30 2021-06-30  122200.0

[880 rows x 2 columns]

In [152]:
# month = df["대여건수"].resample("M").sum()

In [254]:
# 대여건수만 Dataframe으로 가져가기
month = df["대여건수"]
month

대여일자
2019-01-01      8601.0
2019-01-02     14360.0
2019-01-03     16344.0
2019-01-04     17308.0
2019-01-05     12380.0
                ...   
2021-06-26    105813.0
2021-06-27     92869.0
2021-06-28    106378.0
2021-06-29    111907.0
2021-06-30    122200.0
Name: 대여건수, Length: 880, dtype: float64

In [255]:
month = month.to_frame()

In [256]:
# 행 인덱스에서 뒤에 일이 나와있는데 이것을 제거하기위한 코드
new_mon = []
for mon in month.index:
    mon = str(mon)
    new_mon.append(re.search("[\d]+-[\d]+",mon).group())

In [257]:
new_mon

['2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-01',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-02',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '2019-03',
 '20

In [258]:
month.index

DatetimeIndex(['2019-01-01', '2019-01-02', '2019-01-03', '2019-01-04',
               '2019-01-05', '2019-01-06', '2019-01-07', '2019-01-08',
               '2019-01-09', '2019-01-10',
               ...
               '2021-06-21', '2021-06-22', '2021-06-23', '2021-06-24',
               '2021-06-25', '2021-06-26', '2021-06-27', '2021-06-28',
               '2021-06-29', '2021-06-30'],
              dtype='datetime64[ns]', name='대여일자', length=880, freq=None)

In [261]:
month.index = new_mon
month

대여건수
2019-01    8601.0
2019-01   14360.0
2019-01   16344.0
2019-01   17308.0
2019-01   12380.0
...           ...
2021-06  105813.0
2021-06   92869.0
2021-06  106378.0
2021-06  111907.0
2021-06  122200.0

[880 rows x 1 columns]

In [232]:
month.to_excel("월별_공공자전거_대여건수.xlsx")